measure

```diff
@@ -0,0 +1,181 @@
+            assert settings, "Config located at {} was malformed; `{}` application component `{}` provided no settings".format(CONFIG_PATH, DRIVER_NAME, c_name)
+
+            for s_name, s_data in settings.items():
+                for s in REQUIRED_SETTINGS_FIELDS:
+                    assert s in s_data, \
+                        "missing required key '{}' from driver configuration for setting {}, component {}".format(
+                            s, s_name, c_name)
+
+        command = config.get('command')
+        assert isinstance(command, str), 'The `command` attribute in `batch` section must be a string, found {}'.format(type(command).__name__)
+
+        ed = config.get('expected_duration')
+        if ed:
+            assert (isinstance(ed, int) or isinstance(ed, float)) and ed > 0, 'The `expected_duration` attribute in `batch` section must be a numeric type greater than zero when specified. Found: {} `{}`'.format(type(ed).__name__, ed)
+
+        metrics = config.get('metrics')
+        if metrics:
+            assert isinstance(metrics, dict), 'The `metrics` attribute in `batch` section must be a dict when provided, found {}'.format(type(metrics).__name__)
+
+            for m_name, m_data in metrics.items():
+                out_reg = m_data.get('output_regex')
+                assert isinstance(out_reg, str), 'The `output_regex` attribute is required in metric `{}` under `batch` section and must be a string, found {}'.format(m_name, type(metrics).__name__)
+                try:
+                    re.compile(out_reg)
+                except re.error as e:
+                    raise Exception('Failed to compile the `output_regex` of metric `{}` in `batch` section'.format(m_name)) from e
+
+    def describe(self):
+        metrics = { 'total_runtime': { 'unit': 'seconds' } }
+        for m_name, m_data in self.config.get('metrics', {}).items():
+            metrics[m_name] = {}
+            unit = m_data.get('unit')
+            if unit:
+                metrics[m_name]['unit'] = unit
+
+        return metrics
+
+    def _print_progress_wrapper(self):
+        if self.start_time:
+            elapsed_time = time.time() - self.start_time
+            self.progress = int(round(elapsed_time / self.config.get('expected_duration', 3600) * 100))
+        
+        super().print_progress()
+
+    def measure(self):
+        try:
+            in_metrics = self.input_data['metrics']
+        except:
+            raise Exception('Measure input is missing "metrics" key')
+
+        control = self.input_data.get('control')
+        state = state_store.get_state()
+        # Construct a state from config defaults if batch adjust has not occurred yet 
+        if not state:
+            def_state_comps = {}
+            for c_name, c_data in self.config["application"]["components"].items():
```

Comment:
The "application" key can contain other things besides "components", currently just "annotations".